internal/generator/ecdsa/template/ecdsa.go.tmpl

```diff
@@ -0,0 +1,287 @@
+}
+
+var zeroReader = zr{}
+
+const (
+	aesIV = "gnark-crypto IV." // must be 16 chars (equal block size)
+)
+
+func nonce(privateKey *PrivateKey, hash []byte) (csprng *cipher.StreamReader, err error) {
+	// This implementation derives the nonce from an AES-CTR CSPRNG keyed by:
+	//
+	//    SHA2-512(privateKey.scalar ∥ entropy ∥ hash)[:32]
+	//
+	// The CSPRNG key is indifferentiable from a random oracle as shown in
+	// [Coron], the AES-CTR stream is indifferentiable from a random oracle
+	// under standard cryptographic assumptions (see [Larsson] for examples).
+	//
+	// [Coron]: https://cs.nyu.edu/~dodis/ps/merkle.pdf
+	// [Larsson]: https://web.archive.org/web/20040719170906/https://www.nada.kth.se/kurser/kth/2D1441/semteo03/lecturenotes/assump.pdf
+
+	// Get 256 bits of entropy from rand.
+	entropy := make([]byte, 32)
+	_, err = io.ReadFull(rand.Reader, entropy)
+	if err != nil {
+		return
+
+	}
+
+	// Initialize an SHA-512 hash context; digest...
+	md := sha512.New()
+	md.Write(privateKey.scalar[:sizeFr]) // the private key,
+	md.Write(entropy)                   // the entropy,
+	md.Write(hash)                      // and the input hash;
+	key := md.Sum(nil)[:32]             // and compute ChopMD-256(SHA-512),
+	// which is an indifferentiable MAC.
+
+	// Create an AES-CTR instance to use as a CSPRNG.
+	block, _ := aes.NewCipher(key)
+
+	// Create a CSPRNG that xors a stream of zeros with
+	// the output of the AES-CTR instance.
+	csprng = &cipher.StreamReader{
+		R: zeroReader,
+		S: cipher.NewCTR(block, []byte(aesIV)),
+	}
+
+	return csprng, err
+}
+
+// Equal compares 2 public keys
+func (pub *PublicKey) Equal(x signature.PublicKey) bool {
+	xx, ok := x.(*PublicKey)
+	if !ok {
+		return false
+	}
+	bpk := pub.Bytes()
+	bxx := xx.Bytes()
+	return subtle.ConstantTimeCompare(bpk, bxx) == 1
+}
+
+// Public returns the public key associated to the private key.
+func (privKey *PrivateKey) Public() signature.PublicKey {
+	var pub PublicKey
+	pub.A.Set(&privKey.PublicKey.A)
+	return &pub
+}
+
+// Sign performs the ECDSA signature
+//
+// k ← 𝔽r (random)
+// P = k ⋅ g1Gen
+// r = x_P (mod order)
+// s = k⁻¹ . (m + sk ⋅ r)
+// signature = {r, s}
+//
+// SEC 1, Version 2.0, Section 4.1.3
+func (privKey *PrivateKey) Sign(message []byte, hFunc hash.Hash) ([]byte, error) {
+	scalar, r, s, kInv := new(big.Int), new(big.Int), new(big.Int), new(big.Int)
+	scalar.SetBytes(privKey.scalar[:sizeFr])
+	for {
+		for {
+			csprng, err := nonce(privKey, message)
+			if err != nil {
+				return nil, err
+			}
+			k, err := randFieldElement(csprng)
+			if err != nil {
+				return nil, err
+			}
+
+			var P {{ .CurvePackage }}.G1Affine
+			P.ScalarMultiplicationBase(k)
+			kInv.ModInverse(k, order)
+
+			P.X.BigInt(r)
+			r.Mod(r, order)
+			if r.Sign() != 0 {
+				break
+			}
+		}
+		s.Mul(r, scalar)
+
+		// compute the hash of the message as an integer
+		dataToHash := make([]byte, len(message))
+		copy(dataToHash[:], message[:])
+		hFunc.Reset()
```

Comment:
signature.Signer interface says that if hFunc is nil then implementation may assume pre-hashed message. I think it would make sense to also consider this case.

Reply:
This is the case of eddsa too. Should we merge this and do a PR for both ecdsa and eddsa?

Reply:
Yep, makes sense.

Reply:
done in `signature/pre-hashed` branch. Will push the small PR when this one is merged.